### PR TITLE
policy: cache aggregated list of selectors in rule

### PR DESCRIPTION
**Pull request by jrajahalme**
Add a new, unexported field to the `IngressRule` and `EgressRule` types in the
API called `aggregatedSelectors`.

For `IngressRule`, it is an aggregation of the following fields within the type,
all converted to `EndpointSelector`:

* `FromEntities`
* `FromCIDR`
* `FromCIDRSet`

For `EgressRule`, it is an aggregation of the following fields within the type,
all converted to `EndpointSelector`:

* `ToEntities`
* `ToCIDR`
* `ToCIDRSet`
* `ToFQDNs`

Previously, upon each policy calculation, for each rule, this aggregate list of
rule fields as `EndpointSelector` was created in an ad-hoc manner, and was not
performant, as it allocated many different slices of `EndpointSelector` in calls
to `GetSourceEndpointSelectorsWithRequirements` and
`GetDestinationEndpointSelectorsWithRequirements`. This new field caches the
aggregated list only once, upon rule sanitization, or in calls to the two
aforementioned functions if the field has not been initialized yet.

The performance gains by making this change are significant. With 1000
egress rules containing a ToCIDR with 2 CIDRs, and 3000 identities
from before making this change:

```
PASS: resolve_test.go:168: PolicyTestSuite.BenchmarkRegeneratePolicyRules	     100	  12630576 ns/op	 4127697 B/op	  106148 allocs/op
```

And after:

```
PASS: resolve_test.go:168: PolicyTestSuite.BenchmarkRegeneratePolicyRules	     500	   6208546 ns/op	 2265445 B/op	   61134 allocs/op
```

This is a 50 % performance improvement in policy calculation for this simple
policy benchmark.

Note: This is a revival of PR #7250 (commit 21fe2cc18c99 of the same name) that got reverted in f64cf1c420f2 ("policy: Remove more dead code.") due to not being used any more.

Signed-off-by: Ian Vernon <ian@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8282)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/api/egress.go [hunk 1]
```diff
@@ -148,12 +148,34 @@ type EgressRule struct {
...
+// SetAggregatedSelectors creates a single slice containing all of the following
+// fields within the EgressRule, converted to EndpointSelector, to be stored
+// within the EgressRule for easy lookup while performing policy evaluation
+// for the rule:
+// * ToEntities
+// * ToCIDR
+// * ToCIDRSet
+// * ToFQDNs
+func (e *EgressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
```
Comment (ianvernon) [inline pkg/policy/api/egress.go]: Add a comment indicating why we are not appending `ToEndpoints` here.

File: pkg/policy/api/ingress.go [hunk 2]
```diff
@@ -108,12 +108,32 @@ type IngressRule struct {
...
+}
+
+// SetAggregatedSelectors creates a single slice containing all of the following
+// fields within the IngressRule, converted to EndpointSelector, to be stored
+// within the IngressRule for easy lookup while performing policy evaluation
+// for the rule:
+// * FromEntities
+// * FromCIDR
+// * FromCIDRSet
+func (i *IngressRule) SetAggregatedSelectors() {
```
Comment (ianvernon) [inline pkg/policy/api/ingress.go:126]: Add a comment indicating why we are not appending `FromEndpoints` here.

**Review by ianvernon — APPROVED** [review]

File: pkg/policy/api/egress.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/policy/api/egress.go]: Added.

File: pkg/policy/api/ingress.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/policy/api/ingress.go:126]: Added.

**Comment by jrajahalme** [comment]
Rebased and addressed Ian's review comments.

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/api/egress.go [hunk 3]
```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
...
+	res = append(res, e.ToFQDNs.GetAsEndpointSelectors()...)
+	e.aggregatedSelectors = res
 }
 
 // GetDestinationEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
 // all L3 source selectors of the ingress rule
 func (e *EgressRule) GetDestinationEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
-	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
+	if e.aggregatedSelectors == nil {
+		e.SetAggregatedSelectors()
```
Comment (aanm) [inline pkg/policy/api/egress.go:182]: There is no protection against parallel usage of `EgressRule` for this to happen safely.

File: pkg/policy/api/ingress.go [hunk 4]
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
...
+	res = append(res, i.FromCIDRSet.GetAsEndpointSelectors()...)
+	i.aggregatedSelectors = res
 }
 
 // GetSourceEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
 // all L3 source selectors of the ingress rule
 func (i *IngressRule) GetSourceEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
-	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
+	if i.aggregatedSelectors == nil {
+		i.SetAggregatedSelectors()
```
Comment (aanm) [inline pkg/policy/api/ingress.go:140]: There is no protection against parallel usage of `IngressRule` for this to happen safely.

File: pkg/policy/api/egress.go [hunk 5]
```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
...
+// for the rule:
+// * ToEntities
+// * ToCIDR
+// * ToCIDRSet
+// * ToFQDNs
+//
+// ToEndpoints is not aggregated due to requirement folding in
+// GetDestinationEndpointSelectorsWithRequirements()
+func (e *EgressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
```
Comment (aanm) [inline pkg/policy/api/egress.go]: If To endpoints is not being used why the `len(e.ToEndpoints)` here?

File: pkg/policy/api/ingress.go [hunk 6]
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
...
+// within the IngressRule for easy lookup while performing policy evaluation
+// for the rule:
+// * FromEntities
+// * FromCIDR
+// * FromCIDRSet
+//
+// FromEndpoints is not aggregated due to requirement folding in
+// GetSourceEndpointSelectorsWithRequirements()
+func (i *IngressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
```
Comment (aanm) [inline pkg/policy/api/ingress.go]: If `FromEndpoints` is not being used why the `len(i.FromEndpoints) here

File: pkg/policy/api/egress.go [hunk 7]
```diff
@@ -148,12 +148,37 @@ type EgressRule struct {
 	//     - 'sg-XXXXXXXXXXXXX'
 	// +optional
 	ToGroups []ToGroups `json:"toGroups,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
```
Comment (aanm) [inline pkg/policy/api/egress.go:153]: We are mixing the API with implementation details. This structure should kept somewhere else.

File: pkg/policy/api/ingress.go [hunk 8]
```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
 	//
 	// +optional
 	FromEntities EntitySlice `json:"fromEntities,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
```
Comment (aanm) [inline pkg/policy/api/ingress.go:113]: Same here for the mix of API vs implementation details.

Comment (aanm) [inline pkg/policy/api/ingress.go:113]: Same here for the mix of API vs implementation details.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/policy/api/egress.go (hunk 5 above)
Comment (jrajahalme) [reply pkg/policy/api/egress.go]: I though I had fixed this, but apparently I've managed to lose those changes. Thanks for noticing!

File: pkg/policy/api/ingress.go (hunk 6 above)
Comment (jrajahalme) [reply pkg/policy/api/ingress.go]: Fixed.

File: pkg/policy/api/egress.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/policy/api/egress.go:182]: It's safe even if multiple goroutines run this in parallel. They will compute the same result so the order in which they set the pointer does not matter. I've added a comment to this effect.

File: pkg/policy/api/ingress.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/policy/api/ingress.go:140]: It's safe even if multiple goroutines run this in parallel. They will compute the same result so the order in which they set the pointer does not matter. I've added a comment to this effect.

File: pkg/policy/api/egress.go (hunk 7 above)
Comment (jrajahalme) [reply pkg/policy/api/egress.go:153]: I fully agree. However, this was here before, and refactoring to an internal `IngressRule` and `EgressRule` is a bit too much for this cycle. Opened a new Issue #8353 to track this. Also added TODO comments here with reference to the GH issue.

File: pkg/policy/api/ingress.go (hunk 8 above)
Comment (jrajahalme) [reply pkg/policy/api/ingress.go:113]: Same.

Comment (jrajahalme) [reply pkg/policy/api/ingress.go:113]: Same.

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
Fix for the recently merged FQDN PR (#8117)

**Comment by jrajahalme** [comment]
test-me-please

File: pkg/policy/api/ingress.go (hunk 4 above)
Comment (aanm) [reply pkg/policy/api/ingress.go:140]: this feels really odd... also, @jrajahalme comment not added :-)

File: pkg/policy/api/egress.go (hunk 3 above)
Comment (aanm) [reply pkg/policy/api/egress.go:182]: this feels really odd... also, @jrajahalme comment not added :-)

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by jrajahalme** [comment]
> On Jun 19, 2019, at 9:52 AM, André Martins <notifications@github.com> wrote:
> 
> @aanm requested changes on this pull request.
> 
> In pkg/policy/api/ingress.go <https://github.com/cilium/cilium/pull/8282#discussion_r295401566>:
> 
> >  }
>  
>  // GetSourceEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
>  // all L3 source selectors of the ingress rule
>  func (i *IngressRule) GetSourceEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
> -	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
> +	if i.aggregatedSelectors == nil {
> +		i.SetAggregatedSelectors()
> this feels really odd... also, @jrajahalme <https://github.com/jrajahalme> comment not added :-)
> 
> 

The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)

> 
> In pkg/policy/api/egress.go <https://github.com/cilium/cilium/pull/8282#discussion_r295401600>:
> 
> >  }
>  
>  // GetDestinationEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
>  // all L3 source selectors of the ingress rule
>  func (e *EgressRule) GetDestinationEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
> -	res := make(EndpointSelectorSlice, 0, len(e.ToCIDR)+len(e.ToFQDNs)+len(e.ToCIDRSet)+len(e.ToEndpoints)+len(e.ToEntities))
> +	if e.aggregatedSelectors == nil {
> +		e.SetAggregatedSelectors()
> this feels really odd... also, @jrajahalme <https://github.com/jrajahalme> comment not added :-)
> 
> 

The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)

File: pkg/policy/api/egress.go (hunk 3 above)
Comment (ianvernon) [reply pkg/policy/api/egress.go:182]: The comment was added in the implementation of `SetAggregatedSelectors`.

Comment (jrajahalme) [reply pkg/policy/api/egress.go:182]: The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)

File: pkg/policy/api/ingress.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/policy/api/ingress.go:140]: The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)

**Pull request merged**